### ISSUE TITLE
HMM with covariates?

### ISSUE THREAD

**Issue by christer-watson** [jmschrei/pomegranate#1094]
More a question for confirmation than anything (and first time posting). Not a bug report.

I am trying to construct a HMM, 2 states with emission distributions that are Gamma distributions. I would like the transition matrices to be dependent on day-of-the-week (and, down-the-line, other variables). I think the term for these would be covariates.

It appears that this is not currently implemented in Pomegranate. I think my only option would be to create a new HMM distribution that had a more complex treatment of transition matrices. Is all that right? If so, it is probably outside my commitment & knowledge level. However, I wanted to confirm there wasn't another approach before giving up.

**Comment by jmschrei** [comment]
Howdy

Yeah, sorry, time-dependent HMMs are super cool but aren't currently supported. If you have a resource for how they're implemented I'd love to take a look to confirm my understanding but I can't commit to implementing them soon.

**Comment by christer-watson** [comment]
Just one last question. Push aside the difficulties of a time covariate. Is something like a temperature covariate (i.e., not cyclic) easy to incorporate, or would that too require creating a new HMM distribution?

**Comment by jmschrei** [comment]
The short answer is you'd probably need to make a new class in either situation but that potentially the modifications would be minimal. 

If these covariates linearly modify the log probabilities in the transition matrix you could modify the  `forward` and `backward` functions (e.g., this line https://github.com/jmschrei/pomegranate/blob/master/pomegranate/hmm/dense_hmm.py#L339 and maybe also https://github.com/jmschrei/pomegranate/blob/master/pomegranate/hmm/dense_hmm.py#L333) to add in your covariates at that time step. Conceptually, if you're able to fix the `forward` and `backward` methods (and pass the tensor of covariates through all the other functions, e.g. `fit` and `forward_backward`) you should also be able to train the model without any other larger modifications.

If these covariates non-linearly modified/generated the transition matrix you'd probably need to do something more complicated that takes in the covariates as input and outputs a transition matrix at each step. I can see this being possible, particularly with PyTorch, but I don't know exactly how you'd optimize the whole thing.

**Comment by christer-watson** [comment]
Thanks for responding here. One last question. Let's say I want to develop a HMM where each hidden state has two different emission distributions. That is, we can observe two different features. One feature is continuous, drawn from a Gaussian, say. The second feature, however, is categorical. Totally different property. 

Is this possible?

Reading through the very-well written documentation, I can't figure out if this is possible or not. I've tried playing around, without any luck.

**Comment by jmschrei** [comment]
Yes, you should be able to use `IndependentComponents`, which treats each feature independently with a different distribution: https://github.com/jmschrei/pomegranate/blob/master/pomegranate/distributions/independent_components.py#L15

**Comment by jonnor** [comment]
IndependentComponents still seems to be missing from the API reference - although there is a bunch of docstrings for it.
